### PR TITLE
Trim examples ahead of time for auto packing

### PR DESCRIPTION
**Pull request by irenedea**
Previously during packing profiling, we were repeatedly trimming the same examples. The longer the trimmed examples are, the more expensive the operation, so for datasets with longer examples (like `sam-mosaic/iv4-chatml-4k`), we saw very noticeable performance impacts.

## Improved Profiling Times
Tested with this [yaml](https://gist.github.com/irenedea/eb646e0b7cb89a380f63ba0ccabf2aa5), changing only the dataset
`mosaicml/dolly_hhrlhf`
* before: 29.977840185165405 s
* after: 23.40160083770752 s

`sam-mosaic/iv4-chatml-4k`
* before: 131.2125198841095 s
* after: 45.155487060546875 s

## Tests
* Manually tested with `mosaicml/dolly_hhrlhf` and `sam-mosaic/iv4-chatml-4k`. Checked that the same packing ratio was selected

### CODE REVIEW

File: llmfoundry/data/packing.py [hunk 1]
```diff
@@ -300,14 +305,18 @@ def auto_packing_ratio(dataloader_cfg: DictConfig,
...
                                         max_ratio, num_packing_ratios,
                                         device_batch_size)
+    
 
     # Obtain the maximum packing_ratio/minimum padding that has no waste.
     # profiling_results are sorted from smallest to largest packing_ratio.
+    n = 0
     packing_ratio = 1
     for packing_ratio_candidate, _, waste in profiling_results:
+        n += 1
```
Comment (irenedea) [inline llmfoundry/data/packing.py:319]:
```suggestion
```

File: llmfoundry/data/packing.py [hunk 2]
```diff
@@ -300,14 +305,18 @@ def auto_packing_ratio(dataloader_cfg: DictConfig,
...
     max_ratio = max_seq_len / 100
+
     profiling_results = profile_packing(dataloader_cfg, tokenizer, min_ratio,
                                         max_ratio, num_packing_ratios,
                                         device_batch_size)
+    
 
     # Obtain the maximum packing_ratio/minimum padding that has no waste.
     # profiling_results are sorted from smallest to largest packing_ratio.
+    n = 0
```
Comment (irenedea) [inline llmfoundry/data/packing.py:316]:
```suggestion
```

File: llmfoundry/data/packing.py [hunk 3]
```diff
@@ -300,14 +305,18 @@ def auto_packing_ratio(dataloader_cfg: DictConfig,
 
     min_ratio = 1
     max_ratio = max_seq_len / 100
+
```
Comment (irenedea) [inline llmfoundry/data/packing.py:308]:
```suggestion
```

File: llmfoundry/data/packing.py [hunk 4]
```diff
@@ -300,14 +305,18 @@ def auto_packing_ratio(dataloader_cfg: DictConfig,
...
     profiling_results = profile_packing(dataloader_cfg, tokenizer, min_ratio,
                                         max_ratio, num_packing_ratios,
                                         device_batch_size)
+
```
Comment (irenedea) [inline llmfoundry/data/packing.py:312]:
```suggestion
```

File: llmfoundry/data/packing.py [hunk 5]
```diff
@@ -61,7 +61,7 @@ def __call__(
             examples: List[Dict[str, torch.Tensor]]) -> Dict[str, torch.Tensor]:
         batch = self.base_collator(examples)
         return self.pack(batch)
-
+
```
Comment (irenedea) [inline llmfoundry/data/packing.py:64]:
```suggestion
```

File: llmfoundry/data/packing.py [hunk 6]
```diff
@@ -385,18 +389,14 @@ def profile_packing(
...
-            if key == 'input_ids':
-                continue
-            for idx, split in enumerate(big_batch[key].split(raw_batch_size)):
-                batches[idx].update({key: split})
-        return batches
+    # # Cut everything down to size
+    sizes, trimmed_examples = _trim_batch(big_batch)
 
     def profile(raw_batch_size: int) -> Tuple[Optional[float], Optional[float]]:
+        # Copy trimmed examples so that the dicts are not shared.
```
Comment (irenedea) [inline llmfoundry/data/packing.py:396]:
```suggestion
        # Copy trimmed examples so that the dicts are not shared between profiling runs.
```

File: llmfoundry/data/packing.py [hunk 7]
```diff
@@ -73,14 +73,13 @@ def pack(self, batch: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
...
         # Cut everything down to size
-        sizes, trimmed_examples = [], []
-        for idx in range(batch['attention_mask'].shape[0]):
-            size, trimmed_example = _extract_trim_batch_idx(batch, idx)
-            sizes.append(size)
-            trimmed_examples.append(trimmed_example)
+        sizes, trimmed_examples = _trim_batch(batch)
+        return self.pack_trimmed_examples(trimmed_examples, sizes)
 
+    def pack_trimmed_examples(self, trimmed_examples: List[Dict[str,
```
Comment (dakinggg) [inline llmfoundry/data/packing.py:80]: while we're here..can you add a docstring?

File: llmfoundry/data/packing.py [hunk 8]
```diff
@@ -102,6 +101,18 @@ def pack(self, batch: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
         return batch
 
 
+def _trim_batch(
```
Comment (dakinggg) [inline llmfoundry/data/packing.py:104]: docstring here would be nice too. took me a minute to figure out that this was trimming padding off of all examples

File: llmfoundry/data/packing.py [hunk 9]
```diff
@@ -385,18 +396,14 @@ def profile_packing(
...
-        input_ids = big_batch['input_ids'].split(raw_batch_size)
-        batches = [{'input_ids': x} for x in input_ids]
-
-        for key in big_batch.keys():
-            if key == 'input_ids':
-                continue
-            for idx, split in enumerate(big_batch[key].split(raw_batch_size)):
-                batches[idx].update({key: split})
-        return batches
+    # # Cut everything down to size
```
Comment (dakinggg) [inline llmfoundry/data/packing.py:399]:
```suggestion
    # Cut everything down to size
```

**Review by dakinggg — APPROVED** [review]
